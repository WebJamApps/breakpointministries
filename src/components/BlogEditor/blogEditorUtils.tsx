import { Editor } from '@tinymce/tinymce-react';
import React from 'react';

const editor = (blogContent: string | undefined, handleEditorChange: (...args:any)=>any): JSX.Element => (
  <Editor
    apiKey={process.env.TINY_KEY}
    value={blogContent}
    init={{
      height: 600,
      menubar: 'insert tools',
      menu: { format: { title: 'Format', items: 'forecolor backcolor' } },
      plugins: [
        'advlist autolink lists link image charmap print preview anchor',
        'searchreplace visualblocks code fullscreen',
        'insertdatetime media table paste code help wordcount',
      ],
      toolbar:
          'undo redo | formatselect | bold italic backcolor forecolor |'
          + 'alignleft aligncenter alignright alignjustify |'
          + 'bullist numlist outdent indent | removeformat | help',
    }}
    onEditorChange={handleEditorChange}
  />
);

export default { editor };
