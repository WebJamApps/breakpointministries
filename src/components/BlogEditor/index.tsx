import React from 'react';
import type { IBlog, Homepage } from '../../containers/Homepage';
import Forms from '../../lib/forms';
import utils from './blogEditorUtils';

type PageProps = { comp:Homepage;editBlog:IBlog; };
const BlogEditor = ({
  comp, editBlog,
}:PageProps): JSX.Element => {
  const { title, body } = editBlog;
  return (
    <div className="page-content">
      <h1 className="admin--header">
        Post A New Blog
      </h1>
      <div className="changeHomepage">
        <form
          id={`update-blog${editBlog._id}`}
          className="update-blog"
          style={{
            textAlign: 'left', marginLeft: '4px', maxWidth: '100%',
          }}
        >
          {Forms.makeInput({
            type: 'text',
            label: 'Blog Title',
            isRequired: false,
            onChange: (evt:React.ChangeEvent<HTMLInputElement>) => {
              const newEditBlog = { ...editBlog, title: evt.target.value };
              comp.setState({ editBlog: newEditBlog });
            },
            value: title,
            width: '280px',
          })}
          <p>&nbsp;</p>
          <label className="contentLabel" htmlFor="content">
            Content
            <br />
            {utils.editor(body, comp.handleEditorChange)}
          </label>
          <div style={{ marginLeft: '10px', marginTop: '10px' }} id="update-blog-button-div">
            <button
              style={{ marginRight: '10px' }}
              type="button"
              id="update-blog-button"
              className="btn"
              disabled={false}
              onClick={() => comp.putAPI()}
            >
              Update Blog
            </button>
            <button
              style={{ marginRight: '10px' }}
              type="button"
              id="cancel-blog-update-button"
              className="btn"
              disabled={false}
              onClick={() => comp.setState({ editBlog: { _id: '', title: '', body: '' } })}
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default BlogEditor;

