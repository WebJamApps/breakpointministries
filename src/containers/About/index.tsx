import React, { RefObject } from 'react';
import { connect } from 'react-redux';
import { withResizeDetector } from 'react-resize-detector';
import commonUtils from '../../lib/commonUtils';
import mapStoreToProps from '../../redux/mapStoreToProps';

type AboutProps = {
  targetRef: RefObject<HTMLDivElement>;
  width: number;
  height: number;
};

export class About extends React.Component<AboutProps> {
  commonUtils: typeof commonUtils;

  parentRef: React.RefObject<unknown>;

  constructor(props: AboutProps) {
    super(props);
    this.commonUtils = commonUtils;
    this.parentRef = React.createRef();
  }

  async componentDidMount(): Promise<void> {
    this.commonUtils.setTitleAndScroll('About', window.screen.width);
  }

  render(): JSX.Element {
    const { width, targetRef } = this.props;
    return (
      <div ref={targetRef}>
        {width >= 900
          ? (
            <div className="page-content">
              <h1 style={{ marginLeft: '1rem' }}>About Me</h1>
              <p style={{ fontSize: '1.6rem', marginLeft: '1rem', marginBottom: '0' }}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                Ut enim ad minim veniam,
                quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit
                in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident,
                sunt in culpa qui officia deserunt mollit anim id est laborum.
              </p>
              &nbsp;
              <p style={{ fontSize: '1.6rem', marginLeft: '1rem', marginBottom: '0' }}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                Ut enim ad minim veniam,
                quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit
                in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident,
                sunt in culpa qui officia deserunt mollit anim id est laborum.
              </p>
              &nbsp;
              <p style={{ fontSize: '1.6rem', marginLeft: '1rem', marginBottom: '0' }}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                Ut enim ad minim veniam,
                quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit
                in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident,
                sunt in culpa qui officia deserunt mollit anim id est laborum.
              </p>
            </div>
          )
          : (
            <div className="page-content">
              <hr />
              <p style={{ fontSize: '.8rem', marginBottom: '0' }}>&nbsp;</p>
              <p style={{ fontSize: '.8rem', marginBottom: '0' }}>&nbsp;</p>
            </div>
          )}
      </div>
    );
  }
}

export default connect(mapStoreToProps, null)(withResizeDetector(About));
