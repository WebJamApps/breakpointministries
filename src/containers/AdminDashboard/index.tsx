import React, { Component, Dispatch } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { connect } from 'react-redux';
import mapStoreToProps, { Ibook } from '../../redux/mapStoreToProps';
import forms from '../../lib/forms';
import AdminController from './AdminController';
import commonUtils from '../../lib/commonUtils';

export interface PicData {
  buttonId: string; buttonClick: (e: React.ChangeEvent<EventTarget>) => Promise<string>; title: string; nameId: string;
}
export interface DashboardProps extends RouteComponentProps {
  dispatch: Dispatch<unknown>;
  homeContent: Ibook;
  auth: { token: string };
  books: Record<string, string>[];
  showTable: boolean;
  editPic: Ibook;
  youthPics: Ibook[];
  familyPics: Ibook[];
  otherPics: Ibook[];
  musicPics: Ibook[];
}
type DashboardState = {
  type: string;
  title: string;
  homePageContent: string;
  announcementtitle: string;
  announcementurl: string;
  youthName: string;
  youthURL: string;
  forumId: string;
  showCaption: string;
  firstEdit: boolean;
};
export class AdminDashboard extends Component<DashboardProps, DashboardState> {
  commonUtils: { setTitleAndScroll: (pageTitle: string, width: number) => void; };

  controller: AdminController;

  forms: typeof forms;

  constructor(props: DashboardProps) {
    super(props);
    this.commonUtils = commonUtils;
    this.controller = new AdminController(this);
    this.state = {
      type: '',
      title: props.homeContent.title || '',
      homePageContent: props.homeContent.comments || '',
      announcementtitle: '',
      announcementurl: '',
      youthName: '',
      youthURL: '',
      forumId: '',
      showCaption: '',
      firstEdit: true,
    };
    this.forms = forms;
    this.onChange = this.onChange.bind(this);
    this.changeHomepage = this.changeHomepage.bind(this);
  }

  componentDidMount(): void { this.commonUtils.setTitleAndScroll('Admin Dashboard', window.screen.width); }

  onChange(evt: React.ChangeEvent<HTMLInputElement>): string {
    evt.persist();
    this.setState((prevState) => ({ ...prevState, [evt.target.id]: evt.target.value, firstEdit: false }));
    return evt.target.id;
  }

  changeHomepage(): JSX.Element {
    const { title, homePageContent } = this.state;
    const inputParams = {
      type: 'text', label: 'Title', isRequired: false, onChange: this.onChange, value: title, width: '280px',
    };
    return (
      <div className="horiz-scroll">
        <div className="material-content elevation3" style={{ width: '100%', margin: '30px auto' }}>
          <form
            id="post-blog"
            style={{
              textAlign: 'left', marginLeft: '4px', width: '100%', maxWidth: '100%',
            }}
          >
            {this.forms.makeInput(inputParams)}
            <p>&nbsp;</p>
            <label htmlFor="content" style={{ fontSize: '12pt', fontWeight: 'bold' }}>
              Content
              <br />
              {this.controller.editor(homePageContent)}
            </label>
            <div style={{ marginLeft: '2px', marginTop: '10px' }}>
              <button className="btn" style={{ padding: '4px' }} type="button" id="c-h" disabled={false} onClick={this.controller.createBlogAPI}>
                Post Blog
              </button>
            </div>
          </form>
        </div>
      </div>
    );
  }

  render(): JSX.Element {
    return (
      <div className="page-content" style={{ margin: 'auto', width: '79vw' }}>
        <h1 style={{ textAlign: 'center', marginTop: '10px' }}>
          Post A New Blog
        </h1>
        {this.changeHomepage()}
      </div>
    );
  }
}

export default withRouter(connect(mapStoreToProps, null)(AdminDashboard));
