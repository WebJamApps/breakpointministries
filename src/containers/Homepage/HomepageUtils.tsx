import type { Homepage } from './index';

const putAPI = async (comp: Homepage): Promise<string> => {
  const { editBlog } = comp.state;
  const { auth } = comp.props;
  let r;
  try {
    r = await comp.superagent.put(`${process.env.BackendUrl}/blog/${editBlog._id}`)
      .set('Authorization', `Bearer ${auth.token}`)
      .set('Accept', 'application/json')
      .send({ body: editBlog.body, title: editBlog.title });
  } catch (e: any) { return `${e.message}`; }
  return comp.finishAPI('update', r);
};

export default { putAPI };