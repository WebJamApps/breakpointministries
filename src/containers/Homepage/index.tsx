import React, { RefObject } from 'react';
import { FacebookShareButton, FacebookIcon, LinkedinShareButton, LinkedinIcon, TwitterShareButton, TwitterIcon } from 'react-share';
import Superagent from 'superagent';
import Moment from 'react-moment';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import ReactHtmlParser from 'react-html-parser';
import { withResizeDetector } from 'react-resize-detector';
import CommonUtils from '../../lib/commonUtils';
import mapStoreToProps from '../../redux/mapStoreToProps';
import BlogEditor from '../../components/BlogEditor';
import DefaultFooter from '../../App/Footer';

export interface IBlog { created_at?: string; _id: string; title: string, body: string }

type HomepageProps = {
  targetRef: RefObject<HTMLDivElement>;
  width: number;
  height: number;
  blogs: any[];
  auth: any;
};

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface HomepageState {
  editBlog: IBlog;
  referrer: string;
}

export class Homepage extends React.Component<HomepageProps, HomepageState> {
  public commonUtils: typeof CommonUtils = CommonUtils;

  public superagent: Superagent.SuperAgentStatic = Superagent;

  parentRef: React.RefObject<unknown>;

  constructor(props: HomepageProps) {
    super(props);
    this.parentRef = React.createRef();
    // eslint-disable-next-line react/no-unused-state
    this.state = { editBlog: { title: '', body: '', _id: '' }, referrer: '' };
    this.handleEditorChange = this.handleEditorChange.bind(this);
  }

  async componentDidMount(): Promise<void> {
    this.commonUtils.setTitleAndScroll('', window.screen.width);
    const params = new URLSearchParams(window.location.search);
    this.checkBlogId(params);
  }

  handleEditorChange(body: string): void {
    const { editBlog } = this.state;
    const newEditBlog = { ...editBlog, body };
    // eslint-disable-next-line react/no-unused-state
    this.setState({ editBlog: newEditBlog });
  }

  // TODO here is example for building the link for the share buttons
  // http://localhost:9000/?id=6043ee1df6a24931fd372290

  // eslint-disable-next-line class-methods-use-this
  checkBlogId(params: URLSearchParams): void {
    const myId = params.get('id');
    if (myId) {
      const blog = document.getElementById(myId);
      if (blog)blog.scrollIntoView();
    }
  }

  // eslint-disable-next-line class-methods-use-this
  finishAPI(type: string, r: Superagent.Response):string {
    if (r.status === 200) {
      window.location.reload();
      return `${r.status}`;
    }
    return `Failed to ${type} blog, ${r.body ? r.body.message : ''}`;
  }

  async deleteBlog(id:string): Promise<string> {
    const { auth } = this.props;
    let r;
    try {
      r = await this.superagent.delete(`${process.env.BackendUrl}/blog/${id}`)
        .set('Authorization', `Bearer ${auth.token}`)
        .set('Accept', 'application/json');
    } catch (e) { return `${e.message}`; }
    return this.finishAPI('delete', r);
  }

  async putAPI(): Promise<string> {
    const { editBlog } = this.state;
    const { auth } = this.props;
    let r;
    try {
      r = await this.superagent.put(`${process.env.BackendUrl}/blog/${editBlog._id}`)
        .set('Authorization', `Bearer ${auth.token}`)
        .set('Accept', 'application/json')
        .send({ body: editBlog.body, title: editBlog.title });
    } catch (e) { return `${e.message}`; }
    return this.finishAPI('update', r);
  }

  makeEditBlogSection(blog: IBlog): void { this.setState({ editBlog: blog }); }

  editBlogButton(blog: IBlog): JSX.Element {
    return (
      <button
        id={`editBlogButton${blog._id}`}
        style={{ width: '50px' }}
        type="button"
        onClick={() => this.makeEditBlogSection(blog)}
      >
        Edit
      </button>
    );
  }

  deleteBlogButton(id: string): JSX.Element {
    return (
      <button
        id={`deleteBlogButton${id}`}
        style={{ width: '50px' }}
        type="button"
        onClick={() => this.deleteBlog(id)}
      >
        Delete
      </button>
    );
  }

  addBlogButton(): JSX.Element {
    return (
      <button type="button" id="addBlogButton" onClick={() => this.setState({ referrer: '/admin#admin-top' })}>
        <i className="fa fa-plus" />
      </button>
    );
  }

  blogEnder(blog: IBlog): JSX.Element {
    let newTime;
    // eslint-disable-next-line prefer-destructuring
    if (blog.created_at !== undefined && blog.created_at !== null) {
      newTime = <Moment format="MM/DD/YYYY hh:mm">{blog.created_at}</Moment>;
    }
    return (
      <div className="blog__ender">
        <div style={{ display: 'inline-block' }}>
          <div className="blog__time-stamp">
            <span>
              Posted: &nbsp;
              {newTime}
            </span>
          </div>
          {// TODO remove process.env check when feature is working
      /* istanbul ignore next */process.env.NODE_ENV !== 'production' ? this.socialMedia(blog._id) : null
          }
        </div>
      </div>
    );
  }

  // eslint-disable-next-line class-methods-use-this
  blogNotFound(): JSX.Element {
    return (
      <div className="blog__entry">
        <section className="blog__entry--body">
          <p className="blog__entry--no-blogs">There are no blog entries at this time.</p>
        </section>
      </div>
    );
  }

  createBlogButtons(blog: IBlog): JSX.Element {
    const { auth } = this.props;
    const valid = auth.isAuthenticated && auth.user && auth.user.userType;
    return (
      <>
        <span className="blog__entry--button-container__add-blog">{valid ? this.addBlogButton() : null}</span>
        <span className="blog__entry--button-container__delete-blog">
          {valid ? this.deleteBlogButton(blog._id) : null}
        </span>
        <span className="blog__entry--button-container__edit-blog">{valid ? this.editBlogButton(blog) : null}</span>
      </>
    );
  }

  makeBlogArticle(): JSX.Element {
    const { targetRef, blogs } = this.props;
    console.log(blogs);
    return (
      <>
        <div className="blog-container">
          <div className="blog" ref={targetRef}>
            {blogs && blogs.length > 0 ? blogs.map((blog) => (
              <div key={`blog_entry${blog._id}`} className="blog__entry">
                <section className="blog__entry--body">
                  <h2 className="blog__entry--header" id={blog._id}>
                    {ReactHtmlParser(blog && blog.title ? blog.title : '')}
                  </h2>
                  <div className="blog__entry--button-container">
                    {this.createBlogButtons(blog)}
                  </div>
                  <div className="blog__entry--paragraph">
                    {ReactHtmlParser(blog && blog.body ? blog.body : '')}
                  </div>
                  {this.blogEnder(blog)}
                </section>
              </div>
            ))
              : (
                this.blogNotFound()
              )}
          </div>
        </div>
        <DefaultFooter />
      </>
    );
  }

  // eslint-disable-next-line class-methods-use-this
  makeLink(id: string, type: string): JSX.Element {
    /* eslint-disable jsx-a11y/anchor-is-valid */
    return (
        <li key={`${type}${id}`}>
          <a href="#" className={`blog__social-media--link ${type}`} aria-label={`Link to ${type} page`}>
            <i key={id} className={`fab fa-${type}`} />
          </a>
        </li>
    );
  }

  facebookButton(id: string) {
    return(
      <FacebookShareButton
        url={process.env.NODE_ENV !== 'production' ? 'google.com' : `${process.env.BackendUrl}/blog/?id=${id}`}
      >
        <FacebookIcon round size={26} />
      </FacebookShareButton>
    )
  }

  twitterButton(id: string) {
    return (
      <TwitterShareButton
        url={process.env.NODE_ENV !== 'production' ? 'google.com' : `${process.env.BackendUrl}/blog/?id=${id}`}
      >
        <TwitterIcon round size={26} />
      </TwitterShareButton>
    )
  }

  linkedinButton(id: string) {
    return(
      <LinkedinShareButton
      url={process.env.NODE_ENV !== 'production' ? 'google.com' : `${process.env.BackendUrl}/blog/?id=${id}`}
      >
        <LinkedinIcon round size={26} />
      </LinkedinShareButton>
    )
  }

  socialMedia(id: string): JSX.Element {
    return (
      <div style={{ display: 'grid' }}>
        <ul className="blog__social-media">
          {this.facebookButton(id)}
          {this.twitterButton(id)}
          {this.linkedinButton(id)}
          <li key={`url${id}`}>
            {/*TODO: Add onClick function to copy url into clipboard, a link if need be after.*/}
            <a key={`urll${id}`} href={`/?id=${id}`} className="blog__social-media--link copylink" aria-label="Permanent link to blog posting">
              <i key={id} className="fas fa-link" />
            </a>
          </li>
        </ul>
      </div>
    );
  }

  render(): JSX.Element {
    const { editBlog, referrer } = this.state;
    if (editBlog._id !== '') { return (<BlogEditor comp={this} editBlog={editBlog} />); }
    if (referrer !== '') return <Redirect to={referrer} />;
    return (this.makeBlogArticle());
  }
}

export default connect(mapStoreToProps, null)(withResizeDetector(Homepage));
