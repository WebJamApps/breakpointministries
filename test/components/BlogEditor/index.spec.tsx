/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import BlogEditor from '../../../src/components/BlogEditor';
import { IBlog } from '../../../src/containers/Homepage';

describe('Dashboard Container', () => {
  let wrapper:ShallowWrapper<any>, compStub: any, editBlog:IBlog;
  beforeEach(() => {
    editBlog = { title: '', _id: '123', body: '' };
    compStub = {
      setState: jest.fn(),
      utils: { putAPI: jest.fn() },
    };
    wrapper = shallow<typeof BlogEditor>(<BlogEditor
      comp={compStub}
      editBlog={editBlog}
    />);
  });
  it('renders snapshot correctly', () => { expect(wrapper).toMatchSnapshot(); });
  it('handles onChange for blog title', () => {
    const myForm = wrapper.find('#update-blog123');
    myForm.find('#blogtitle').simulate('change', { target: { value: 'howdy' } });
    expect(compStub.setState).toHaveBeenCalledWith({ editBlog: { title: 'howdy', _id: '123', body: '' } });
  });
  it('handles click for update-blog-button', () => {
    const myForm = wrapper.find('#update-blog123');
    const buttonDiv = myForm.find('#update-blog-button-div');
    buttonDiv.find('#update-blog-button').simulate('click');
    expect(compStub.utils.putAPI).toHaveBeenCalled();
  });
  it('handles click for cancel-blog-update-button', () => {
    const myForm = wrapper.find('#update-blog123');
    const buttonDiv = myForm.find('#update-blog-button-div');
    buttonDiv.find('#cancel-blog-update-button').simulate('click');
    expect(compStub.setState).toHaveBeenCalled();
  });
});
