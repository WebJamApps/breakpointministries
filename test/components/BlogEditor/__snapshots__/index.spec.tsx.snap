// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Container renders snapshot correctly 1`] = `ShallowWrapper {}`;
